atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/transaction/impl/AbstractSerializableTransactionTest.java

```diff
@@ -1408,6 +1414,47 @@ public void testGetSortedColumnsConflictForReadBoundaryEntry() {
         readSortedColumnsAndInduceConflict(100, 100, true);
     }
 
+    @Test
+    public void testShouldNotReadValuesAtSuccessorTimestamp() {
+        TransactionManager manager = createManager();
+
+        manager.runTaskThrowOnConflict(tx -> {
+            tx.put(TEST_TABLE_SERIALIZABLE, ImmutableMap.of(CELL_ONE, BYTES_ONE));
+            return null;
+        });
+
+        assertThatThrownBy(() -> manager.runTaskThrowOnConflict(tx -> {
+            long myTs = tx.getTimestamp();
+            tx.get(TEST_TABLE_SERIALIZABLE, ImmutableSet.of(CELL_ONE));
+
+            // This write is done to ensure that the transaction actually has writes that need to commit
+            tx.put(TEST_TABLE_SERIALIZABLE, ImmutableMap.of(CELL_TWO, BYTES_TWO));
+
+            keyValueService.put(TEST_TABLE_SERIALIZABLE, ImmutableMap.of(CELL_ONE, BYTES_TWO), myTs + 1);
+            transactionService.putUnlessExists(myTs + 1, myTs + 2);
```

Comment:
we should comment that this is effectively orchestrating a "B" transaction (if this is "ABA").

Reply:
Yeah maybe a bit more comment on what we are doing